### ISSUE TITLE
Concurrent requests produce DPI-1002: invalid dpiLob handle 

### ISSUE THREAD

**Issue by amaru0601** [godror/godror#304]
**Describe the bug**
When executing a stored procedure that returns a clob from an Oracle database using the godror library, the procedure works correctly when run sequentially. However, when making multiple concurrent requests, an error occurs.

**To Reproduce**
1. Set up a stored procedure in Oracle that returns a clob.
2. Use the godror library to execute the procedure sequentially. (This should work correctly).
3. Try making multiple concurrent requests to the same procedure.
4. Observe the error.

```go

const qry = `DECLARE
	VXMLIN    CLOB;
	VXMLOUT   CLOB;
 BEGIN
	VXMLIN := :vxmlin;

	VXMLOUT := NULL;

	TC.SP_RETURNBIG_DATA (VXMLIN, VXMLOUT);

	:2 := VXMLOUT;
	COMMIT;
 END;
`
var vxmlin godror.Lob = godror.Lob{}
var clobResult godror.Lob = godror.Lob{IsClob: true}

stmt, err := db.PrepareContext(ctx, query)
if err != nil {
    return fmt.Errorf("%s: %w", query, err)
}
defer stmt.Close()

if _, err := stmt.ExecContext(ctx, sql.Named("vxmlin", vxmlin),sql.Out{Dest: &clobResult, In: true}); err != nil {
	log.Printf("Error running %q: %+v", query, err)
	return
}

if clobResult.Reader != nil {
		bodyBytes, err := io.ReadAll(clobResult.Reader). // here the error occurs
		if err != nil {
			return nil, fmt.Errorf("error al convertir a bytes la respuesta: %v", err)
		}
		bodyString := string(bodyBytes)

```

**Expected behavior**
I expected the stored procedure to execute correctly and return the clob, even when making multiple concurrent requests.

**Error output**
getChunkSize: ORA-00000: DPI-1002: invalid dpiLob handle

**Your oracle client version**
12.2.0.1.0

**Your godror version**
v0.37.0

**Your go version**
1.20

**Machine (please complete the following information):**
 - OS: win
 - Architecture: x86_64
 - Version: 10

**Additional context**
I am using gin and I externally execute a go script to send multiple request at same time. Also I tried with standalone true and false for both stages same error. Here the setup of the db:
```go
        standAloneConn := false
        poolSize := 16
        var connString string
        if standAloneConn {
            connString = fmt.Sprintf("oracle://%s:%s@%s%s?connClass=POOLED standaloneConnection=1 poolMinSessions=2 poolMaxSessions=16 poolIncrement=2 poolSessionMaxLifetime=300s poolSessionTimeout=60s poolWaitTimeout=300s", user, password, host+":"+port, dbName)
        } else {
            connString = fmt.Sprintf("oracle://%s:%s@%s%s?connClass=POOLED standaloneConnection=0 poolMinSessions=2 poolMaxSessions=16 poolIncrement=2 poolSessionMaxLifetime=300s poolSessionTimeout=60s poolWaitTimeout=300s", user, password, host+":"+port, dbName)
        }

        dbConnO, err = sql.Open(driverName, connString)
        if standAloneConn {
            dbConnO.SetMaxIdleConns(poolSize / 2)
            dbConnO.SetConnMaxLifetime(10 * time.Minute)
        } else {
            dbConnO.SetConnMaxLifetime(0)
        }
        dbConnO.SetMaxOpenConns(poolSize)
        if err != nil {
            return nil, nil, errors.New(err.Error()) 
        }
        err = dbConnO.Ping()
        if err != nil {
            return nil, nil, errors.New(err.Error())
        }

        
```

**Comment by tgulacsi** [comment]
Yes, LOBs are Conn-specific.Either use a Conn per goroutine, or a Tx.

**Comment by amaru0601** [comment]
Thank you for your response. If I understand correctly:

In simpler terms, you're suggesting that I should:
- Establish a separate database connection for each concurrent operation (request or goroutine) I'm running, or
- Group my operations within a database transaction to ensure they're executed in a controlled and consistent manner.

I'll look into refactoring my code based on your advice. If I encounter any challenges, I'll reach out for further clarification.

**Comment by tgulacsi** [comment]
Yes, LOBs are tied to the connection they're returned from the database.
________________________________

**Comment by amaru0601** [comment]
Thank you for your advice! I refactored my code to use a transaction and the issue was resolved. By the way, is there any documentation, warning, or reference regarding this specific scenario with LOBs? I'd like to understand more about this...

**Comment by amaru0601** [comment]
@tgulacsi  what happens if my database has not configured DRCP and I want to use standalone in false?... my connection would die and it wouldn't reconnect, right?

**Comment by tgulacsi** [comment]
No DRCP and standalone=false means session pooling.
If your connection dies, then your session closes, the transaction is rolled back (when the server notices the connection loss).
But this is the same for DRCP, too (maybe longer threshold for deciding a connection is lost).

**Comment by amaru0601** [comment]
and no DRCP and standalone=true means session pooling also?

**Comment by tgulacsi** [comment]
No, godror does not pool sessions.
But Go database/sql does.
If you don't want session pooling, set db.SetMaxIdleConns(0).


________________________________

**Comment by amaru0601** [comment]
oh nice @tgulacsi for your answer! So it means that database/sql does a pool of sessions... is there no difference between connection and session for database/sql? I was reading about oracle documentation they talk about logical and physical connection but for this topic I understand that database/sql connection and session are the same...

**Comment by tgulacsi** [comment]
database/sql is an abstraction over what a database/sql/driver exposes as a DriverConn.
Here, godror exposes an Oracle session/connection (the parlance depends on history and whether the connection is pooled either in the OCI library or in the database, or neither.

So a session can be cached in the database (DRCP), InstantClient (pooled), database/sql (*sql.DB itself is a connection pool), or in your application (i.e. keeping Txs around till finishing off).


________________________________
